### DISCUSSION TITLE
Monorepo with babel aliases support

### DISCUSSION THREAD

**Discussion by yuriy-yarosh** [avajs/ava#3264] [Q&A]
I'm using ava with `babel-preset-typescript` and a set of path aliases for monorepo, e.g. `@org/testing-common/src/Common.ts` to `testing-common/Common.ts` aliasing.

`ava.config.cjs`
```js
module.exports = {
  files: [
    './test/**/*Test.ts',
  ],
  require: [
    '@org/config-testing/Register.js',
  ],
  snapshotDir: 'test/snapshots',
  concurrency: 64,
  failFast: true,
  failWithoutAssertions: false,
  verbose: true,
  nodeArguments: [
    '--trace-deprecation',
  ],
  nodeArguments: [
    '--experimental-modules'
  ],
  extensions: ['ts', 'tsx', 'mts']
};
``` 
Both `@ava/typescript` and plain babel register results with the same error.

Register.js
```js
const babelRC = require('./.babelrc.js');

require('@babel/register')({
  extensions: ['.es6', '.es', '.jsx', '.js', '.mjs', '.ts', 'tsx', '.mts'],
  cache: false,
  ...babelRC,
});
```

`.babelrc` with a custom expo preset and an alias resolver.
```
const { production } = require('@org/config-babel-preset-expo/Resolver.js')

module.exports = {
  "env": {
    "coverage": {
      "plugins": ["istanbul"]
    }
  },
  "presets": [
    "@babel/preset-typescript",
    "@org/config-babel-preset-expo",
    ["@babel/preset-env", {
      "forceAllTransforms": false,
      "modules": "auto",
      "targets": production
    }],
    ["@babel/preset-react", {
      "runtime": "automatic",
    }],
  ],
  "plugins": [
    "babel-plugin-transform-typescript-metadata",
    "macros",
    ["@babel/plugin-proposal-decorators", { "legacy": true }],
    ["@babel/plugin-proposal-class-properties", { "loose": true  }]
  ]
}
```

With aliased imports for tests 
```js
import { hasDir, isFile } from 'config-testing-common/Common';
```

I'm always getting 
```
 export * from 'config-testing-common/Mocked';
  ^^^^^^

  SyntaxError: Unexpected token 'export'
```

But relative imports for my ava test files work just fine
```js
import { hasDir, isFile } from '../src/Common';
```
Everything else everywhere is aliased.

My ava config doesn't work without a custom  `@babel/register` - using `@ava/typescript` config only gives
```
SyntaxError: Cannot use import statement outside a module
```

**Comment by yuriy-yarosh** [comment]
At least I've noticed that aliased imports are working correctly only for the current package.
I can't import anything across yarn workspace without prior compilation with babel and the respective path rewriting, which is puzzling. 

Maybe ava.js has something similar to `get-package-type` somewhere, similarly [to tape](https://github.com/ljharb/tape/blob/master/bin/import-or-require.js#L11), so the respective rewrites module aliases could be forced with a custom config there (e.g. to specify explicitly that an alias is a module).

**Reply by novemberborn** [reply]
You can do that by file type: https://github.com/avajs/ava/blob/main/docs/06-configuration.md#configuring-module-formats

All these loaders make things very complicated. AVA does the right thing with `require()` or `import()` when loading your test file, and from there on the test file controls how things are loaded. My advice is to compile to disk and rewrite the import, which is what `@ava/typescript` does.

**Reply by yuriy-yarosh** [reply]
@novemberborn right now, the only way to make ava work with TypeScript this way (on Node 20), is to use TSX loader via `NODE_OPTIONS='--import tsx' yarn ava`.

The `ts-node/esm` loader import via NODE_OPTIONS as `NODE_OPTIONS='--import ts-node/esm yarn ava`, hangs up completely, for unknown reason.

It would be great if someone could figure something out with `ts-node`.
tsx has wayyy too many limitations e.g. no decorators support for TypeORM / MikroORM in esbuild https://github.com/evanw/esbuild/issues/104
I'd rather use swc instead of esbuild there.

**Reply by novemberborn** [reply]
This is why `@ava/typescript` exists — it avoids the entire mess by having the compiler be external to AVA.
